src/tests/test-visitor.ts

```diff
@@ -38,23 +38,117 @@ describe('VariablesTrackingVisitor', () => {
+  const visit$include = (includeString: string) =>
+    visitor.visitNode(new yaml.$include(includeString), 'test', testEnv)
+
+  it('visit$include with no dots', function() {
+    expect(visit$include('bar')).to.deep.equal($env.bar)
+  });
+
+  it('visit$include with dots', function() {
+    expect(visit$include('obj.0')).to.deep.equal($env.obj[0])
+    expect(visit$include('obj.a')).to.deep.equal($env.obj.a);
+    expect(visit$include('obj.b.c')).to.deep.equal($env.obj.b.c);
+    expect(visit$include('obj.b.d')).to.deep.equal($env.obj.b.d);
+    expect(visit$include('obj.b.d.e')).to.deep.equal($env.obj.b.d.e);
+    expect(visit$include('obj.b.d.f')).to.deep.equal($env.obj.b.d.f);
+    expect(visit$include('obj.b.d.f.g')).to.deep.equal($env.obj.b.d.f.g);
+  });
+
+  it('visit$include with dynamic sub-keys', function() {
+    expect(visit$include('obj[keyA]')).to.deep.equal($env.obj.a)
+    expect(visit$include('obj[key0]')).to.deep.equal($env.obj[0])
+    expect(visit$include('obj[0]')).to.deep.equal($env.obj[0])
+    expect(visit$include('obj[keyB]')).to.deep.equal($env.obj.b)
+    expect(visit$include('obj[keyB].c')).to.deep.equal($env.obj.b.c)
+    expect(visit$include('obj.keyA')).to.deep.equal($env.obj.keyA)
+    expect(visit$include('obj[obj.keyA]')).to.deep.equal($env.obj.a)
+    expect(visit$include('obj[obj.keyB].c')).to.deep.equal($env.obj.b.c)
+    expect(visit$include('obj.keys[keyB]')).to.deep.equal('b')
+  });
+
+  it('visit$include with multiple dynamic sub-keys', function() {
+    expect(visit$include('obj[keyB][keyC]')).to.deep.equal($env.obj.b.c)
+    expect(visit$include('obj[keyB][obj.keyC]')).to.deep.equal($env.obj.b.c)
+    expect(visit$include('obj[obj.keyB][obj.keyC]')).to.deep.equal($env.obj.b.c)
+    expect(visit$include('obj[obj.keys[keyB]].c')).to.deep.equal($env.obj.b.c)
+    expect(visit$include('obj[keyB][obj.keyD].e')).to.deep.equal($env.obj.b.d.e)
+    expect(visit$include('obj.b[obj[keyB].keyC]')).to.deep.equal($env.obj.b.c)
+    expect(visit$include('obj.b[obj[keyB].keyD].e')).to.deep.equal($env.obj.b.d.e)
+    expect(visit$include('obj.b[obj[keyB].keyD][keyE]')).to.deep.equal($env.obj.b.d.e)
+  });
+
+});
+
+describe('_parse$includeKeyChunks', () => {
+  it('all examples parse correctly', () => {
+    const examples = {
+      'foo': ['foo'],
+      'foo[sub]': ['foo[sub]'],
+      'foo.bar': ['foo', 'bar'],
+      'foo.bar.baz': ['foo', 'bar', 'baz'],
+      'foo[sub].bar': ['foo[sub]', 'bar'],
+      'foo[sub].bar.baz': ['foo[sub]', 'bar', 'baz'],
+      'foo[sub].bar.baz[sub2]': ['foo[sub]', 'bar', 'baz[sub2]'],
+      'foo.bar[sub].baz': ['foo', 'bar[sub]', 'baz'],
+      'foo[sub0.sub1].bar': ['foo[sub0.sub1]', 'bar'],
+      'foo[sub0][sub1].bar': ['foo[sub0][sub1]', 'bar'],
+      'foo[sub0][sub1].bar[sub3]': ['foo[sub0][sub1]', 'bar[sub3]']
+    }
+    for (const [key, res] of Object.entries(examples)) {
+      expect(_parse$includeKeyChunks(key, 'test'), `input=${key}`).to.deep.equal(res);
```

Comment:
Not sure if I missed something obvious but it took me way too long to discover what the second argument to `expect` is for - a custom error message. TIL.

Reply:
Yep, custom message.